<a href="https://colab.research.google.com/github/GrantPennington/CS430_InClassAssignment2_GTP/blob/main/InClassAssignment2_GTP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h3>Grant Pennington</h3>
<h4>CS-430</h4>
<h4>In Class Assignment 2</h4>

## Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/InClassAssignment2/newhousing.csv')

## Analyze Dataset

In [3]:
data.head()

price  area  bedrooms  bathrooms  stories  mainroad  guestroom  basement  \
0  5250000  5500         3          2        1         1          0         1   
1  4480000  4040         3          1        2         1          0         0   
2  3570000  3640         2          1        1         1          0         0   
3  2870000  3040         2          1        1         0          0         0   
4  3570000  4500         2          1        1         0          0         0   

   hotwaterheating  airconditioning  parking  prefarea  semi-furnished  \
0                0                0        0         0               1   
1                0                0        1         0               1   
2                0                0        0         0               0   
3                0                0        0         0               0   
4                0                0        0         0               0   

   unfurnished  areaperbedroom   bbratio  
0            0     1833.333333  0.666667  
1            0     1346.666667  0.333333  
2            1     1820.000000  0.500000  
3            1     1520.000000  0.500000  
4            0     2250.000000  0.500000

In [4]:
data.describe()

price          area    bedrooms   bathrooms     stories  \
count  5.450000e+02    545.000000  545.000000  545.000000  545.000000   
mean   4.766729e+06   5150.541284    2.965138    1.286239    1.805505   
std    1.870440e+06   2170.141023    0.738064    0.502470    0.867492   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.430000e+06   3600.000000    2.000000    1.000000    1.000000   
50%    4.340000e+06   4600.000000    3.000000    1.000000    2.000000   
75%    5.740000e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

         mainroad   guestroom    basement  hotwaterheating  airconditioning  \
count  545.000000  545.000000  545.000000       545.000000       545.000000   
mean     0.858716    0.177982    0.350459         0.045872         0.315596   
std      0.348635    0.382849    0.477552         0.209399         0.465180   
min      0.000000    0.000000    0.000000         0.000000         0.000000   
25%      1.000000    0.000000    0.000000         0.000000         0.000000   
50%      1.000000    0.000000    0.000000         0.000000         0.000000   
75%      1.000000    0.000000    1.000000         0.000000         1.000000   
max      1.000000    1.000000    1.000000         1.000000         1.000000   

          parking    prefarea  semi-furnished  unfurnished  areaperbedroom  \
count  545.000000  545.000000      545.000000   545.000000      545.000000   
mean     0.693578    0.234862        0.416514     0.326606     1819.852599   
std      0.861586    0.424302        0.493434     0.469402      839.091825   
min      0.000000    0.000000        0.000000     0.000000      381.000000   
25%      0.000000    0.000000        0.000000     0.000000     1237.500000   
50%      0.000000    0.000000        0.000000     0.000000     1666.666667   
75%      1.000000    0.000000        1.000000     1.000000     2183.333333   
max      3.000000    1.000000        1.000000     1.000000     6600.000000   

          bbratio  
count  545.000000  
mean     0.446361  
std      0.159492  
min      0.166667  
25%      0.333333  
50%      0.400000  
75%      0.500000  
max      1.000000

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   price            545 non-null    int64  
 1   area             545 non-null    int64  
 2   bedrooms         545 non-null    int64  
 3   bathrooms        545 non-null    int64  
 4   stories          545 non-null    int64  
 5   mainroad         545 non-null    int64  
 6   guestroom        545 non-null    int64  
 7   basement         545 non-null    int64  
 8   hotwaterheating  545 non-null    int64  
 9   airconditioning  545 non-null    int64  
 10  parking          545 non-null    int64  
 11  prefarea         545 non-null    int64  
 12  semi-furnished   545 non-null    int64  
 13  unfurnished      545 non-null    int64  
 14  areaperbedroom   545 non-null    float64
 15  bbratio          545 non-null    float64
dtypes: float64(2), int64(14)
memory usage: 68.2 KB


## Split the Data

### Target column is price

In [6]:
X = data.drop('price', axis=1)
y = data[['price']]

## Feature Scaling

### Scale X and y data

In [7]:
from sklearn.preprocessing import StandardScaler
sc_X=StandardScaler()
sc_y=StandardScaler()
X_scaled=sc_X.fit_transform(X)
y_scaled=sc_y.fit_transform(y)

## Train Test Split

In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=42, shuffle=True)

## Linear Regression

In [9]:
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

LinearRegression()

In [10]:
lin_reg.score(X_train, y_train)

0.6760953066606615

In [11]:
lin_reg.score(X_test, y_test)

0.6814896416143947

## Random Forest

In [12]:
from sklearn.ensemble import RandomForestRegressor
rf = RandomForestRegressor()
rf.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


RandomForestRegressor()

In [13]:
rf.score(X_train, y_train)

0.9385055169701255

In [14]:
rf.score(X_test, y_test)

0.7008472214013177

## PCA

In [15]:
from sklearn.decomposition import PCA
pca=PCA(n_components=5)
pca.fit(X_scaled)
X_pca=pca.transform(X_scaled)

In [16]:
X_pca.shape

(545, 5)

In [17]:
X_pca

array([[ 0.6608588 , -0.17104646,  0.29030255,  2.36977878,  0.54755181],
       [-0.75750595, -0.45325575, -1.24020117,  0.47830892, -1.28436511],
       [-1.91995217,  1.54975914,  1.24787064, -0.1021777 ,  0.10205909],
       ...,
       [-0.09605631,  1.65985667, -0.49436167, -0.53234478, -0.89040693],
       [ 2.54323222,  4.34098798,  0.0854805 , -0.71476077, -1.04341683],
       [-1.94832471, -0.90014852, -0.83130394,  0.96820513, -0.84611608]])

## PCA With LinearRegression

### Train test split again with PCA using X_pca and y_scaled

In [18]:
X_train, X_test, y_train, y_test = train_test_split(X_pca, y_scaled, test_size = 0.3, random_state=42, shuffle=True)

In [19]:
lin_reg_pca = LinearRegression()
lin_reg_pca.fit(X_train, y_train)

LinearRegression()

In [20]:
lin_reg_pca.score(X_train, y_train)

0.6524055085553088

In [21]:
lin_reg_pca.score(X_test, y_test)

0.6814004558937791

## PCA With Random Forest

In [22]:
rf_pca = RandomForestRegressor()
rf_pca.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


RandomForestRegressor()

In [23]:
rf_pca.score(X_train, y_train)

0.9383407430089176

In [24]:
rf_pca.score(X_test, y_test)

0.6746368216452014

## My test scores for RandomForest and PCA RandomForest were a lot lower than the training scores so I know the model is overfitting. 


## Because of this, the best model to use is <b>LinearRegression</b> with test score of 0.681